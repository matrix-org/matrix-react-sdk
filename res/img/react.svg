<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
