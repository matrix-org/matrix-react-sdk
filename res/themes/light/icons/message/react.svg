<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>