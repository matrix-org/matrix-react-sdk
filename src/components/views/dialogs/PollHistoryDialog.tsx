/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { MatrixClient, Room } from "matrix-js-sdk/src/matrix";

import { RoomPermalinkCreator } from "../../../utils/permalinks/Permalinks";
import { PollHistory } from "../polls/pollHistory/PollHistory";
import BaseDialog from "./BaseDialog";

type PollHistoryDialogProps = {
    room: Room;
    matrixClient: MatrixClient;
    permalinkCreator: RoomPermalinkCreator;
    onFinished(): void;
};

export const PollHistoryDialog: React.FC<PollHistoryDialogProps> = ({
    room,
    matrixClient,
    permalinkCreator,
    onFinished,
}) => {
    // @TODO hide dialog title somehow
    return (
        <BaseDialog onFinished={onFinished}>
            <PollHistory
                room={room}
                matrixClient={matrixClient}
                permalinkCreator={permalinkCreator}
                onFinished={onFinished}
            />
        </BaseDialog>
    );
};
