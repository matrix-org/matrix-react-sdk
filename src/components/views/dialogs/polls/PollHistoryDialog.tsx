/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { useState } from "react";
import { MatrixClient } from "matrix-js-sdk/src/client";
import { MatrixEvent, Poll } from "matrix-js-sdk/src/matrix";

import { _t } from "../../../../languageHandler";
import BaseDialog from "../BaseDialog";
import { IDialogProps } from "../IDialogProps";
import { PollHistoryList } from "./PollHistoryList";
import { PollHistoryFilter } from "./types";
import { usePollsWithRelations } from "./usePollHistory";
import { useFetchPastPolls } from "./fetchPastPolls";

type PollHistoryDialogProps = Pick<IDialogProps, "onFinished"> & {
    roomId: string;
    matrixClient: MatrixClient;
};

const sortEventsByLatest = (left: MatrixEvent, right: MatrixEvent): number => right.getTs() - left.getTs();
const filterPolls =
    (filter: PollHistoryFilter) =>
    (poll: Poll): boolean =>
        // exclude polls while they are still loading
        // to avoid jitter in list
        !poll.isFetchingResponses && (filter === "ACTIVE") !== poll.isEnded;

const filterAndSortPolls = (polls: Map<string, Poll>, filter: PollHistoryFilter): MatrixEvent[] => {
    return [...polls.values()]
        .filter(filterPolls(filter))
        .map((poll) => poll.rootEvent)
        .sort(sortEventsByLatest);
};

export const PollHistoryDialog: React.FC<PollHistoryDialogProps> = ({ roomId, matrixClient, onFinished }) => {
    const room = matrixClient.getRoom(roomId)!;
    const { isLoading } = useFetchPastPolls(room, matrixClient);
    const { polls } = usePollsWithRelations(roomId, matrixClient);
    const [filter, setFilter] = useState<PollHistoryFilter>("ACTIVE");

    const pollStartEvents = filterAndSortPolls(polls, filter);
    const isLoadingPollResponses = [...polls.values()].some((poll) => poll.isFetchingResponses);

    return (
        <BaseDialog title={_t("Polls history")} onFinished={onFinished}>
            <div className="mx_PollHistoryDialog_content">
                <PollHistoryList
                    pollStartEvents={pollStartEvents}
                    isLoading={isLoading || isLoadingPollResponses}
                    polls={polls}
                    filter={filter}
                    onFilterChange={setFilter}
                />
            </div>
        </BaseDialog>
    );
};
