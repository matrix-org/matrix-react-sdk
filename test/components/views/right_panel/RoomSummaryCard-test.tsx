/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { MatrixEvent, Room } from "matrix-js-sdk/src/matrix";

import DMRoomMap from "../../../../src/utils/DMRoomMap";
import RoomSummaryCard from "../../../../src/components/views/right_panel/RoomSummaryCard";
import ShareDialog from "../../../../src/components/views/dialogs/ShareDialog";
import ExportDialog from "../../../../src/components/views/dialogs/ExportDialog";
import MatrixClientContext from "../../../../src/contexts/MatrixClientContext";
import defaultDispatcher from "../../../../src/dispatcher/dispatcher";
import * as settingsHooks from "../../../../src/hooks/useSettings";
import Modal from "../../../../src/Modal";
import RightPanelStore from "../../../../src/stores/right-panel/RightPanelStore";
import { RightPanelPhases } from "../../../../src/stores/right-panel/RightPanelStorePhases";
import { getMockClientWithEventEmitter, mockClientMethodsUser } from "../../../test-utils";
import { PollHistoryDialog } from "../../../../src/components/views/dialogs/polls/PollHistoryDialog";

describe("<RoomSummaryCard />", () => {
    const userId = "@alice:domain.org";
    const mockClient = getMockClientWithEventEmitter({
        ...mockClientMethodsUser(userId),
        isRoomEncrypted: jest.fn(),
        getRoom: jest.fn(),
    });
    const roomId = "!room:domain.org";
    const room = new Room(roomId, mockClient, userId);
    const roomCreateEvent = new MatrixEvent({
        type: "m.room.create",
        room_id: roomId,
        sender: userId,
        content: {
            creator: userId,
            room_version: "5",
        },
        state_key: "",
    });
    room.currentState.setStateEvents([roomCreateEvent]);
    const defaultProps = {
        room,
        onClose: jest.fn(),
    };
    const getComponent = (props = {}) =>
        render(<RoomSummaryCard {...defaultProps} {...props} />, {
            wrapper: ({ children }) => (
                <MatrixClientContext.Provider value={mockClient}>{children}</MatrixClientContext.Provider>
            ),
        });

    const modalSpy = jest.spyOn(Modal, "createDialog");
    const dispatchSpy = jest.spyOn(defaultDispatcher, "dispatch");
    const rightPanelCardSpy = jest.spyOn(RightPanelStore.instance, "pushCard");
    const featureEnabledSpy = jest.spyOn(settingsHooks, "useFeatureEnabled");

    beforeEach(() => {
        jest.clearAllMocks();
        DMRoomMap.makeShared();

        mockClient.getRoom.mockReturnValue(room);
        jest.spyOn(room, "isElementVideoRoom").mockRestore();
        jest.spyOn(room, "isCallRoom").mockRestore();
        featureEnabledSpy.mockReset().mockReturnValue(false);
    });

    it("renders the room summary", () => {
        const { container } = getComponent();
        expect(container).toMatchSnapshot();
    });

    it("opens room members list on button click", () => {
        const { getByText } = getComponent();

        fireEvent.click(getByText("People"));

        expect(rightPanelCardSpy).toHaveBeenCalledWith({ phase: RightPanelPhases.RoomMemberList }, true);
    });

    it("opens room file panel on button click", () => {
        const { getByText } = getComponent();

        fireEvent.click(getByText("Files"));

        expect(rightPanelCardSpy).toHaveBeenCalledWith({ phase: RightPanelPhases.FilePanel }, true);
    });

    it("opens room export dialog on button click", () => {
        const { getByText } = getComponent();

        fireEvent.click(getByText("Export chat"));

        expect(modalSpy).toHaveBeenCalledWith(ExportDialog, { room });
    });

    it("opens share room dialog on button click", () => {
        const { getByText } = getComponent();

        fireEvent.click(getByText("Share room"));

        expect(modalSpy).toHaveBeenCalledWith(ShareDialog, { target: room });
    });

    it("opens room settings on button click", () => {
        const { getByText } = getComponent();

        fireEvent.click(getByText("Room settings"));

        expect(dispatchSpy).toHaveBeenCalledWith({ action: "open_room_settings" });
    });

    describe("pinning", () => {
        it("renders pins options when pinning feature is enabled", () => {
            featureEnabledSpy.mockImplementation((feature) => feature === "feature_pinning");
            const { getByText } = getComponent();

            expect(getByText("Pinned")).toBeInTheDocument();
        });
    });

    describe("poll history", () => {
        it("renders poll history option when feature is enabled", () => {
            featureEnabledSpy.mockImplementation((feature) => feature === "feature_poll_history");
            const { getByText } = getComponent();

            expect(getByText("Polls history")).toBeInTheDocument();
        });

        it("opens poll history dialog on button click", () => {
            featureEnabledSpy.mockImplementation((feature) => feature === "feature_poll_history");
            const { getByText } = getComponent();

            fireEvent.click(getByText("Polls history"));

            expect(modalSpy).toHaveBeenCalledWith(PollHistoryDialog, { roomId, matrixClient: mockClient });
        });
    });

    describe("video rooms", () => {
        it("does not render irrelevant options for element video room", () => {
            jest.spyOn(room, "isElementVideoRoom").mockReturnValue(true);
            featureEnabledSpy.mockImplementation(
                (feature) => feature === "feature_video_rooms" || feature === "feature_pinning",
            );
            const { queryByText } = getComponent();

            // options not rendered
            expect(queryByText("Files")).not.toBeInTheDocument();
            expect(queryByText("Pinned")).not.toBeInTheDocument();
            expect(queryByText("Export chat")).not.toBeInTheDocument();
        });

        it("does not render irrelevant options for element call room", () => {
            jest.spyOn(room, "isCallRoom").mockReturnValue(true);
            featureEnabledSpy.mockImplementation(
                (feature) =>
                    feature === "feature_element_call_video_rooms" ||
                    feature === "feature_video_rooms" ||
                    feature === "feature_pinning",
            );
            const { queryByText } = getComponent();

            // options not rendered
            expect(queryByText("Files")).not.toBeInTheDocument();
            expect(queryByText("Pinned")).not.toBeInTheDocument();
            expect(queryByText("Export chat")).not.toBeInTheDocument();
        });
    });
});
