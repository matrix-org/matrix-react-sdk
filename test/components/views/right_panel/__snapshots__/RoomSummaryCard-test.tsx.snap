// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomSummaryCard /> renders the room summary 1`] = `
<div>
  <div
    class="mx_BaseCard mx_RoomSummaryCard"
  >
    <div
      class="mx_AutoHideScrollbar"
      tabindex="-1"
    >
      <header
        class="mx_Flex mx_RoomSummaryCard_header"
        style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: space-between; --mx-flex-gap: var(--cpd-space-3x);"
      >
        <button
          aria-label="Search"
          class="mx_RoomSummaryCard_searchBtn"
          data-state="closed"
          data-testid="summary-search"
        >
          <div
            height="100%"
            width="100%"
          />
        </button>
        <div
          class="mx_AccessibleButton mx_BaseCard_close"
          data-testid="base-card-close-button"
          role="button"
          tabindex="0"
          title="Close"
        />
      </header>
      <header
        class="mx_RoomSummaryCard_container"
      >
        <span
          class="_avatar_1o69u_17 mx_BaseAvatar _avatar-imageless_1o69u_60"
          data-color="7"
          data-testid="avatar-img"
          data-type="round"
          role="presentation"
          style="--cpd-avatar-size: 80px;"
        >
          !
        </span>
        <h1
          class="_typography_yh5dq_162 _font-heading-md-semibold_yh5dq_121 mx_RoomSummaryCard_roomName text-primary"
          title="!room:domain.org"
        >
          !room:domain.org
        </h1>
        <div
          class="_typography_yh5dq_162 _font-body-sm-semibold_yh5dq_45 mx_RoomSummaryCard_alias text-secondary"
          title=""
        />
        <section
          class="mx_Flex mx_RoomSummaryCard_badges"
          style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: start; --mx-flex-justify: center; --mx-flex-gap: var(--cpd-space-2x);"
        >
          <span
            class="_typography_yh5dq_162 _font-body-sm-medium_yh5dq_50 _badge_qipht_17"
            data-kind="default"
          >
            <div
              width="1em"
            />
            Not encrypted
          </span>
        </section>
      </header>
      <div
        class="_separator_1uqhh_17"
        data-orientation="horizontal"
        role="separator"
      />
      <label
        class="_item_zxa40_17 _interactive_zxa40_36"
        data-kind="primary"
        for=":r1:"
        role="menuitemcheckbox"
      >
        <div
          aria-hidden="true"
          class="_icon_zxa40_49"
          height="24"
          width="24"
        />
        <span
          class="_typography_yh5dq_162 _font-body-md-medium_yh5dq_69 _label_zxa40_58"
        >
          Favourite
        </span>
        <div
          class="_container_ik1u1_18"
        >
          <input
            class="_input_ik1u1_32"
            id=":r1:"
            type="checkbox"
          />
          <div
            class="_ui_ik1u1_42"
          />
        </div>
      </label>
      <button
        class="_item_zxa40_17 _interactive_zxa40_36 _disabled_zxa40_125"
        data-kind="primary"
        disabled=""
        role="menuitem"
      >
        <div
          aria-hidden="true"
          class="_icon_zxa40_49"
          height="24"
          width="24"
        />
        <span
          class="_typography_yh5dq_162 _font-body-md-medium_yh5dq_69 _label_zxa40_58"
        >
          Invite
        </span>
        <svg
          aria-hidden="true"
          class="_nav-hint_zxa40_65"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="_item_zxa40_17 _interactive_zxa40_36"
        data-kind="primary"
        role="menuitem"
      >
        <div
          aria-hidden="true"
          class="_icon_zxa40_49"
          height="24"
          width="24"
        />
        <span
          class="_typography_yh5dq_162 _font-body-md-medium_yh5dq_69 _label_zxa40_58"
        >
          Copy link
        </span>
        <svg
          aria-hidden="true"
          class="_nav-hint_zxa40_65"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="_item_zxa40_17 _interactive_zxa40_36"
        data-kind="primary"
        role="menuitem"
      >
        <div
          aria-hidden="true"
          class="_icon_zxa40_49"
          height="24"
          width="24"
        />
        <span
          class="_typography_yh5dq_162 _font-body-md-medium_yh5dq_69 _label_zxa40_58"
        >
          Settings
        </span>
        <svg
          aria-hidden="true"
          class="_nav-hint_zxa40_65"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="_separator_1uqhh_17"
        data-orientation="horizontal"
        role="separator"
      />
      <button
        class="_item_zxa40_17 _interactive_zxa40_36"
        data-kind="primary"
        role="menuitem"
      >
        <div
          aria-hidden="true"
          class="_icon_zxa40_49"
          height="24"
          width="24"
        />
        <span
          class="_typography_yh5dq_162 _font-body-md-medium_yh5dq_69 _label_zxa40_58"
        >
          People
        </span>
        <svg
          aria-hidden="true"
          class="_nav-hint_zxa40_65"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="_item_zxa40_17 _interactive_zxa40_36"
        data-kind="primary"
        role="menuitem"
      >
        <div
          aria-hidden="true"
          class="_icon_zxa40_49"
          height="24"
          width="24"
        />
        <span
          class="_typography_yh5dq_162 _font-body-md-medium_yh5dq_69 _label_zxa40_58"
        >
          Files
        </span>
        <svg
          aria-hidden="true"
          class="_nav-hint_zxa40_65"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="_item_zxa40_17 _interactive_zxa40_36"
        data-kind="primary"
        role="menuitem"
      >
        <div
          aria-hidden="true"
          class="_icon_zxa40_49"
          height="24"
          width="24"
        />
        <span
          class="_typography_yh5dq_162 _font-body-md-medium_yh5dq_69 _label_zxa40_58"
        >
          Poll history
        </span>
        <svg
          aria-hidden="true"
          class="_nav-hint_zxa40_65"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="_item_zxa40_17 _interactive_zxa40_36"
        data-kind="primary"
        role="menuitem"
      >
        <div
          aria-hidden="true"
          class="_icon_zxa40_49"
          height="24"
          width="24"
        />
        <span
          class="_typography_yh5dq_162 _font-body-md-medium_yh5dq_69 _label_zxa40_58"
        >
          Export Chat
        </span>
        <svg
          aria-hidden="true"
          class="_nav-hint_zxa40_65"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="_separator_1uqhh_17"
        data-orientation="horizontal"
        role="separator"
      />
      <button
        class="_item_zxa40_17 _interactive_zxa40_36"
        data-kind="critical"
        role="menuitem"
      >
        <div
          aria-hidden="true"
          class="_icon_zxa40_49"
          height="24"
          width="24"
        />
        <span
          class="_typography_yh5dq_162 _font-body-md-medium_yh5dq_69 _label_zxa40_58"
        >
          Leave room
        </span>
        <svg
          aria-hidden="true"
          class="_nav-hint_zxa40_65"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.70005 17.3C8.51672 17.1167 8.42505 16.8834 8.42505 16.6C8.42505 16.3167 8.51672 16.0834 8.70005 15.9L12.6 12L8.70005 8.10005C8.51672 7.91672 8.42505 7.68338 8.42505 7.40005C8.42505 7.11672 8.51672 6.88338 8.70005 6.70005C8.88338 6.51672 9.11672 6.42505 9.40005 6.42505C9.68338 6.42505 9.91672 6.51672 10.1 6.70005L14.7 11.3C14.8 11.4 14.8709 11.5084 14.9125 11.625C14.9542 11.7417 14.975 11.8667 14.975 12C14.975 12.1334 14.9542 12.2584 14.9125 12.375C14.8709 12.4917 14.8 12.6 14.7 12.7L10.1 17.3C9.91672 17.4834 9.68338 17.575 9.40005 17.575C9.11672 17.575 8.88338 17.4834 8.70005 17.3Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="mx_BaseCard_Group mx_RoomSummaryCard_appsGroup"
      >
        <h2>
          Widgets
        </h2>
        <div
          class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_link"
          role="button"
          tabindex="0"
        >
          Add widgets, bridges & bots
        </div>
      </div>
    </div>
  </div>
</div>
`;
